pkg/visitors/tightenme.go

```diff
@@ -0,0 +1,121 @@
+// Copyright 2019 the LinuxBoot Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package visitors
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+
+	"github.com/linuxboot/fiano/pkg/uefi"
+)
+
+// TightenME tighten ME's belt to give more room for LinuxBoot
+type TightenME struct {
+	fd  *uefi.FlashDescriptor
+	mer *uefi.MERegion
+	br  *uefi.BIOSRegion
+}
+
+// Run wraps Visit and performs some setup and teardown tasks.
+func (v *TightenME) Run(f uefi.Firmware) error {
+	err := f.Apply(v)
+	if err != nil {
+		return fmt.Errorf("error looking for IFD, ME and BIOS regions: %v", err)
+	}
+
+	return v.process()
+}
+
+// Visit applies the TightenME visitor to any Firmware type.
+func (v *TightenME) Visit(f uefi.Firmware) error {
+	switch f := f.(type) {
+	case *uefi.FlashDescriptor:
+		v.fd = f
+		return nil
+	case *uefi.MERegion:
+		v.mer = f
+		return nil
+	case *uefi.BIOSRegion:
+		v.br = f
+		return nil
+	default:
+		return f.ApplyChildren(v)
+	}
+}
+
+func (v *TightenME) process() error {
+	// Ensuring IFD exists also ensure FRegion are populated in Regions
+	if v.fd == nil {
+		return fmt.Errorf("no IFD found")
+	}
+	if v.mer == nil {
+		return fmt.Errorf("no ME region found")
+	}
+	if v.br == nil {
+		return fmt.Errorf("no BIOS region found")
+	}
+	if v.mer.FRegion.EndOffset() != v.br.FRegion.BaseOffset() {
+		return fmt.Errorf("ME and BIOS regions are not contiguous: ME end at %#x BIOS starts at %#x", v.mer.FRegion.EndOffset(), v.br.FRegion.BaseOffset())
+
+	}
+	// Compute the new limit
+	updateOffset := uint64(v.mer.FRegion.BaseOffset()) + v.mer.FreeSpaceOffset
+	updateBase := (updateOffset + uefi.RegionBlockSize - 1) / uefi.RegionBlockSize
+	updateOffset = updateBase * uefi.RegionBlockSize
+	bufOffset := updateOffset - uint64(v.mer.FRegion.BaseOffset())
+	// check the zone if empty
+	buf := v.mer.Buf()
+	if !uefi.IsErased(buf[bufOffset:], uefi.Attributes.ErasePolarity) {
+		return fmt.Errorf("ME unused space in not erased as expected")
+	}
+	// Shrink ME Region
+	v.mer.FRegion.Limit = uint16(updateBase - 1)
+	v.mer.SetBuf(buf[:bufOffset])
+	// Expand BIOS Region
+	offsetShift := uint64(v.br.FRegion.BaseOffset()) - updateOffset
+	fmt.Printf("JVDG %#x %#x %#x\n", v.br.FRegion.BaseOffset(), updateOffset, offsetShift)
+	v.br.FRegion.Base = uint16(updateBase)
+	v.br.Length += offsetShift
```

Comment:
If I remember correctly, there is a restriction in the SPI controller where it can only map at most 16MiB of the BIOS region into memory. I'm not sure if UEFI depends on the whole BIOS region being mapped into memory, but you should probably print a region if it is >16MiB.

Reply:
done